### ISSUE TITLE
Win32 exception when executing a 'write' section

### ISSUE THREAD

**Issue by kikish** [Microsoft/CNTK#223]
Hi,

I try to modelize a problem which have 300 inputs and one output.
The training dataset contains 250 uples and I have a testing dataset that contains 100k uples.

I have 2 sections in the configuration file : one for the training process and a second for saving the predictions of the model on a testing dataset ('write' section).

The training process works well but the 'write' process on the testing dataset raises a Win32 exception. I have the same error if I use the training dataset.
_"Allocating matrices for forward and/or backward propagation.
UCIFastReader: Starting at epoch 0, counting lines to determine record count...
 250 records found.
starting epoch 0 at record count 0, and file position 0
already there from last epoch
RandomOrdering: 44 retries for 250 elements (17.6%) to ensure window condition
RandomOrdering: recached sequence for seed 0: 22, 29, ...
CNTK: Win32 exception caught (such an access violation or a stack overflow)"_

The error message above is the one returned when I use the training dataset in the write section. 

Have you an idea about this issue ?

Regards,
Arezki.

**Comment by mradmila** [comment]
Hi Arezki,

This could be a number of different issues, depending on which Win32 exception is being thrown. You could enable "first chance" exceptions and catch this under a debugger, and then share the callstack with us. We will work on our end to improve the diagnostics in the situations where Win32 exceptions are thrown.

**Comment by kikish** [comment]
Hi,

Ok. I will try that and I will come back soon with the requested callstack.

Thanks.

**Comment by kikish** [comment]
Hi,

Here is the callstack I get from Visual Studio 2013.
======================
 	KernelBase.dll!000007fefd28b3dd()	Inconnu
 	msvcr120d.dll!000007fed64fc366()	Inconnu
>	CNTK.exe!Microsoft::MSR::CNTK::ThrowFormatted<std::runtime_error>(const char * format, ...) Ligne 65	C++
 	CNTK.exe!Microsoft::MSR::CNTK::RuntimeError<wchar_t const * __ptr64,char * __ptr64>(const char * format, const wchar_t * && <_Args_0>, char * && <_Args_1>) Ligne 79	C++
 	CNTK.exe!fopenOrDie(const std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t> > & pathname, const wchar_t * mode) Ligne 253	C++
 	CNTK.exe!Microsoft::MSR::CNTK::File::Init::__l39::<lambda>() Ligne 118	C++
 	CNTK.exe!Microsoft::MSR::CNTK::attempt<void <lambda>(void) >(int retries, const Microsoft::MSR::CNTK::File::Init::__l39::void <lambda>(void) & body) Ligne 65	C++
 	CNTK.exe!Microsoft::MSR::CNTK::attempt<void <lambda>(void) >(const Microsoft::MSR::CNTK::File::Init::__l39::void <lambda>(void) & body) Ligne 93	C++
 	CNTK.exe!Microsoft::MSR::CNTK::File::Init(const wchar_t * filename, int fileOptions) Ligne 120	C++
 	CNTK.exe!Microsoft::MSR::CNTK::File::File(const std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t> > & filename, int fileOptions) Ligne 32	C++
 	CNTK.exe!Microsoft::MSR::CNTK::ConfigParser::ReadConfigFile(const std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t> > & filePath) Ligne 200	C++
 	CNTK.exe!Microsoft::MSR::CNTK::ConfigParser::LoadConfigFileAndResolveVariables(const std::basic_string<wchar_t,std::char_traits<wchar_t>,std::allocator<wchar_t> > & filePath, const Microsoft::MSR::CNTK::ConfigParameters & config) Ligne 156	C++
 	CNTK.exe!Microsoft::MSR::CNTK::NDLBuilder<float>::Init(const Microsoft::MSR::CNTK::ConfigParameters & config) Ligne 149	C++
 	CNTK.exe!Microsoft::MSR::CNTK::NDLBuilder<float>::NDLBuilder<float>(const Microsoft::MSR::CNTK::ConfigParameters & config) Ligne 41	C++
 	[Code externe]	
 	CNTK.exe!DoTrain<Microsoft::MSR::CNTK::ConfigParameters,float>(const Microsoft::MSR::CNTK::ConfigParameters & config) Ligne 123	C++
 	CNTK.exe!DoCommands<float>(const Microsoft::MSR::CNTK::ConfigParameters & config) Ligne 217	C++
 	CNTK.exe!wmainOldCNTKConfig(int argc, wchar_t * * argv) Ligne 589	C++
 	CNTK.exe!wmain1(int argc, wchar_t * * argv) Ligne 629	C++
 	CNTK.exe!wmain(int argc, wchar_t * * argv) Ligne 668	C++
=======================================================

If you want more info, please tell me.

Arezki.

**Comment by kikish** [comment]
Any new information on this issue ?

Regards.

Arezki

**Comment by frankseide** [comment]
Thanks. I am a bit puzzled. You see a Win32 exception, but the call stack clearly says that it did throw a C++ exception. So somehow the C++ exception got masked by a Win32 exception. If we could just see the actual error that ThrowFormatted() throws, we could solve this easily.

I see French in your log, so one very remote possibility is that this has to do with a failed character-code conversion. Are you using non-ASCII characters in the pathnames of any of your NDL files (e.g. vowels with accents)? Could you try pathnames without?

Some theory what might have happened: The call stack contains this line: 
```
                    ndlScript.LoadConfigFileAndResolveVariables(msra::strfun::utf16(filePath), config);
```
where filePath is converted from an 8-bit string to a 16-bit Unicode string. This conversion assumes that filePath is an 8-bit Unicode string. If your config files are Latin-1 encoded (that might be the default on European-language Windows machines, I believe), and filePath contains Latin-1 characters, then the conversion above will be wrong, and fopen() will fail.

But we should see a proper error message. The only reason why we may not see it that I can think of is that the the error string itself is an 8-bit string again. Maybe that back conversion caught an invalid character?

**Comment by frankseide** [comment]
I just noticed that your Win32 error message is from an old version. Could you try the latest? That should give us a proper Win32 error code. Meanwhile, I am pinging the C++ CRT team to see whether the above could be an explanation.

**Comment by kikish** [comment]
Hi,

Thanks for your reply. I have uploaded the last version of CNTK source code that I compiled on my computer. The issue is still there.

I don't use special characters (e.g. voyels) either in path name or in file. I have 2 projects : one with a datset (filse size about 500Mb) used in the write section that contains about 100k lines without the issue (I have my file with the predictions from my NN) and the second (file size about 1Gb) where the issue arise that contains about 350k lines.

Here is the write bloc in case I have wroten something wrong.

write = [
	action = "write"
	minibatchSize = 16
	
	NDLNetworkBuilder=[
        networkDescription = "$ConfigDir$/01_OneHidden_Regr.ndl"
    ]
	
	epochSize = 0
	modelPath = $modelPath$
	#outputNodeNames = [OutputNodes]
	outputPath = "$ModelDir$/output_prediction.txt"
	
    reader = [
        readerType = "UCIFastReader"
        file = "$DataDir$/$etdFileName$"
         customDelimiter="\t"
        features = [
            dim = 452
            start = 0
        ]
        
        labels = [
			labelType = "regression" ## regression, category, none
            dim = 1
            start = 452
			#allowMapCreation = 1
			#labelDim=100250
			labelMappingFile = "$DataDir$/labelsmap.txt"
        ]
    ]
]


The call stacks remains the same. Only the line numbers in the source code change.

Any idea ?

Arezki.

**Comment by frankseide** [comment]
Thanks.

Would you mind trying one thing: In Source/Common/Include/Basics.h, ThrowFormatted(), please change this line
```
    char buffer[bufferSize];
```
to
```
    char buffer[bufferSize+1] = { 0 };
```
and see if the problem goes away. I found that the vsnprintf() function used on this buffer does not seem to behave as documented.

**Comment by kikish** [comment]
Hi,

I found the error. It's my fault. When I defined the write section, I defined a "labels" section within my reader but my dataset doesn't have an output column. Then CNTK try to access to a member variable 'm_labelData' (UCIFastReader.cpp) which is empty then throw an exception.

May be can you add a check to see that the dataset has his number of columns equal to the sum of the dim values defined in the 'features' and 'labels' section ? 

Thanks for your helps.

Regards,

Arezki.

**Comment by frankseide** [comment]
Thanks for letting us know. It is our fault if we don't show you a proper error message. I apologize for the lost time.
